src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/DefaultRazorFormattingService.cs

```diff
@@ -54,9 +65,90 @@ internal class DefaultRazorFormattingService : RazorFormattingService
             _server = server;
             _csharpFormatter = new CSharpFormatter(documentMappingService, server, filePathNormalizer);
             _htmlFormatter = new HtmlFormatter(server, filePathNormalizer);
+            _optionsMonitor = optionsMonitor;
             _logger = loggerFactory.CreateLogger<DefaultRazorFormattingService>();
         }
 
+        public override Task<TextEdit[]> FormatOnTypeAsync(Uri uri, RazorCodeDocument codeDocument, Position position, string character, FormattingOptions options)
+        {
+            if (uri is null)
+            {
+                throw new ArgumentNullException(nameof(uri));
+            }
+
+            if (codeDocument is null)
+            {
+                throw new ArgumentNullException(nameof(codeDocument));
+            }
+
+            if (position is null)
+            {
+                throw new ArgumentNullException(nameof(position));
+            }
+
+            if (string.IsNullOrEmpty(character))
+            {
+                throw new ArgumentNullException(nameof(character));
+            }
+
+            if (options is null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (!_optionsMonitor.CurrentValue.AutoClosingTags || character != ">")
```

Comment:
Do we need this extra check if `AutoClosingTags` here? Wondering if it's sufficient to control this at the Endpoint layer.

Reply:
The thing is we'll eventually remove that check from the endpoint layer. Because imagine we want to support onTypeFormatting for C#. Then we will call into the same API. We can't short circuit it at the endpoint layer.

Reply:
ah i see ok 👍